File: Tools/ClientConfig/tools/xls-to-proto/configLoaderTemplate.cpp
#include "configLoaderTemplate.h"

configLoaderTemplate* configLoaderTemplate::m_pInstance = NULL;

configLoaderTemplate::configLoaderTemplate()
{

}

configLoaderTemplate* configLoaderTemplate::getInstance()
{
	if (NULL == m_pInstance)
	{
		m_pInstance = new configLoaderTemplate();
	}

	return m_pInstance;
}

void configLoaderTemplate::load(const unsigned char* buffer, int bufferLen)
{
	if (NULL == buffer)
	{
		return;
	}

	if (bufferLen <= 4)
	{
		return;
	}

	const unsigned char* begin = buffer;
	const unsigned char* end = buffer + bufferLen;

	while (begin != end)
	{
		int configLength  = 0;
		memcpy(&configLength, begin, 4);
		begin += 4;

		TemplateConfig* config = new TemplateConfig();
		config->ParseFromArray(begin, configLength);
		m_configVectorCache.push_back(config);
		m_configMapCache[config->key()] = config;

		begin += configLength;
	}
}

TemplateConfig* configLoaderTemplate::getConfigByKey(TemplateKey key)
{
	std::map<TemplateKey, TemplateConfig*>::iterator it = m_configMapCache.find(key);

	if (it == m_configMapCache.end())
	{
		return NULL;
	}

	return it->second;
}

std::vector<TemplateConfig*>& configLoaderTemplate::getAllCachedConfig()
{
	return m_configVectorCache;
}

void configLoaderTemplate::releaseConfig()
{
	m_configMapCache.clear();

	for (int i = 0; i < m_configVectorCache.size(); i++)
	{
		TemplateConfig* config = m_configVectorCache[i];
		delete config;
	}

	m_configVectorCache.clear();
}

File: Tools/ClientConfig/tools/xls-to-proto/configLoaderTemplate.cs
using System;
using System.Collections.Generic;
using System.Collections;
using ProtoBuf;
using System.IO;
using Configuration;

class configLoaderTemplate
{
    private configLoaderTemplate()
    {
        m_configCache = new List<TemplateConfig>();
//        m_configHashCache = new Hashtable();
    }

    private static configLoaderTemplate m_instance = null;

    private List<TemplateConfig> m_configCache = null;

//    private Hashtable m_configHashCache = null;

    public static configLoaderTemplate getInstance()
    {
        if (null == m_instance)
        {
            m_instance = new configLoaderTemplate();
        }

        return m_instance;
    }

    public void load(string path)
    {
        if (false == File.Exists(path))
        {
            return;
        }

        byte[] byteAll = File.ReadAllBytes(path);

        if (byteAll == null  || byteAll.Length <= 0)
        {
            return;
        }

        releaseConfig();
           
        int length = BitConverter.ToInt32(byteAll, 0);

        int offset = 4;

        while (offset <= byteAll.Length)
        {
            MemoryStream memStream = new MemoryStream(byteAll, offset, length);

            TemplateConfig config = Serializer.Deserialize<TemplateConfig>(memStream);

            m_configCache.Add(config);

//            m_configHashCache.Add(TemplateKey, config);

            offset += length;

            if (offset >= byteAll.Length)
            {
                break;
            }

            length = BitConverter.ToInt32(byteAll, offset);
            offset += 4;
        }
    }

    public void load(byte[] buffer)
    {
        if (null == buffer || buffer.Length <= 0)
        {
            return;
        }

        releaseConfig();

        int length = BitConverter.ToInt32(buffer, 0);


        int offset = 4;

        while (offset <= buffer.Length)
        {
            MemoryStream memStream = new MemoryStream(buffer, offset, length);

            TemplateConfig config = Serializer.Deserialize<TemplateConfig>(memStream);

            m_configCache.Add(config);

//            m_configHashCache.Add(TemplateKey, config);

            offset += length;

            if (offset >= buffer.Length)
            {
                break;
            }

            length = BitConverter.ToInt32(buffer, offset);
            offset += 4;
        }
    }

 /*   public TemplateConfig getConfigByKey(object key)
    {
        if (null == m_configHashCache)
        {
            return null;
        }

        if (false == m_configHashCache.Contains(key))
        {
            return null;
        }

        return (TemplateConfig)m_configHashCache[key];
    }
*/
	public List<TemplateConfig> getAllCachedConfig()
	{
		return m_configCache;
	}

    public void releaseConfig(){
        m_configCache.Clear();
//        m_configHashCache.Clear();
    }

}

